File: typescript/src/p_0002_add_two_numbers.ts
import * as assert from 'node:assert';

/**
 * Definition for singly-linked list.
 */
class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.next = (next===undefined ? null : next)
    }
}

function addTwoNumbers(l1: ListNode | null, l2: ListNode | null): ListNode | null {
    let dn = new ListNode(0);
    let curr = dn;
    let lr = 0
    while (l1 !== null || l2 !== null || lr !== 0) {
        let l1Val = l1 !== null ? l1.val : 0;
        let l2Val = l2 !== null ? l2.val : 0;
        let sum = lr + l1Val + l2Val;
        curr.next = new ListNode(sum % 10);
        curr = curr.next;
        lr = Math.floor(sum / 10);
        l1 = l1 !== null ? l1.next : null;
        l2 = l2 !== null ? l2.next : null;
    }
    return dn.next;
};

function nodeToInt(node: ListNode | null): number {
    let num = 0;
    let base = 1;
    while (node !== null) {
        num += node.val * base;
        base *= 10;
        node = node.next;
    }
    return num;
}

// Example usage:

let l1 = new ListNode(2, new ListNode(4, new ListNode(3)));
let l2 = new ListNode(5, new ListNode(6, new ListNode(4)));
assert.strictEqual(nodeToInt(addTwoNumbers(l1, l2)), nodeToInt(l1) + nodeToInt(l2));
console.log(`Test case 1 passed: ${nodeToInt(l1)} + ${nodeToInt(l2)} = ${nodeToInt(addTwoNumbers(l1, l2))}`);

l1 = new ListNode(0);
l2 = new ListNode(0);
assert.strictEqual(nodeToInt(addTwoNumbers(l1, l2)), nodeToInt(l1) + nodeToInt(l2));
console.log(`Test case 2 passed: ${nodeToInt(l1)} + ${nodeToInt(l2)} = ${nodeToInt(addTwoNumbers(l1, l2))}`);

l1 = new ListNode(9, new ListNode(9, new ListNode(9, new ListNode(9, new ListNode(9, new ListNode(9, new ListNode(9)))))));
l2 = new ListNode(9, new ListNode(9, new ListNode(9, new ListNode(9))));
assert.strictEqual(nodeToInt(addTwoNumbers(l1, l2)), nodeToInt(l1) + nodeToInt(l2));
console.log(`Test case 3 passed: ${nodeToInt(l1)} + ${nodeToInt(l2)} = ${nodeToInt(addTwoNumbers(l1, l2))}`);

File: python/src/p_0002_add_two_numbers.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val: int = 0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        result = ListNode(0)
        curr = result
        lr = 0
        while l1 is not None or l2 is not None or lr > 0:
            sum = l1.val if l1 else 0
            sum += l2.val if l2 else 0
            sum += lr
            curr.next = ListNode(sum % 10)
            curr = curr.next
            lr = sum // 10
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None

        return result.next


def node_to_int(node: ListNode | None) -> int:
    n = node
    result = 0
    m = 1
    while n is not None:
        result += n.val * m
        m *= 10
        n = n.next
    return result


if __name__ == "__main__":
    s = Solution()
    l1 = ListNode(2, ListNode(4, ListNode(3)))
    assert node_to_int(l1) == 342
    l2 = ListNode(5, ListNode(6, ListNode(4)))
    assert node_to_int(l2) == 465
    r = s.addTwoNumbers(l1, l2)
    print(f'{node_to_int(l1)} + {node_to_int(l2)} = {node_to_int(r)}')
    assert node_to_int(l1) + node_to_int(l2) == node_to_int(r)

    s = Solution()
    l1 = ListNode(2, ListNode(4))
    l2 = ListNode(5, ListNode(6, ListNode(4)))
    r = s.addTwoNumbers(l1, l2)
    print(f'{node_to_int(l1)} + {node_to_int(l2)} = {node_to_int(r)}')
    assert node_to_int(l1) + node_to_int(l2) == node_to_int(r)

    s = Solution()
    l1 = ListNode(9, ListNode(9, ListNode(9, ListNode(9, ListNode(9, ListNode(9, ListNode(9)))))))
    l2 = ListNode(9, ListNode(9, ListNode(9, ListNode(9))))
    r = s.addTwoNumbers(l1, l2)
    print(f'{node_to_int(l1)} + {node_to_int(l2)} = {node_to_int(r)}')
    assert node_to_int(l1) + node_to_int(l2) == node_to_int(r)
    